core/main/src/service/apps/apps_updater.rs

```diff
@@ -11,22 +15,127 @@ use ripple_sdk::{
         },
         extn_client_message::{ExtnMessage, ExtnPayload, ExtnResponse},
     },
+    log::info,
     tokio::sync::mpsc::{Receiver, Sender},
 };
 
 #[cfg(not(test))]
 use ripple_sdk::log::{debug, error};
+use serde::{Deserialize, Serialize};
 
+use std::{
+    fs,
+    sync::{Arc, RwLock},
+};
 #[cfg(test)]
 use std::{println as debug, println as error};
 
 use ripple_sdk::extn::client::extn_processor::ExtnStreamer;
 
+use crate::state::platform_state::PlatformState;
+
 #[derive(Clone)]
 pub struct AppsUpdaterState {
     client: ExtnClient,
     ignore_list: Vec<String>,
     uninstalls_enabled: bool,
+    sideload_support: bool,
+    failed_app_installs: Arc<RwLock<Vec<FailedAppInstall>>>,
```

Comment:
Any reason this is an array instead of a map with the app_id as the key? Then you wouldn't have to do all that iterating later.

Reply:
yeah, it could be. I tend to just lean towards arrays instead of maps so it doesn't lock you into a primary key in that map. Like if we wanted to store multiple failed installed versions for the same app, then we would have to redo the keys. Probably not a big deal, but also think the iter methods are pretty easy to do a lookup by any field